### ISSUE TITLE
Problem with WidgetCondition Combine

### ISSUE THREAD

**Issue by template-provider** [Smile-SA/elasticsuite#2452]
Returning only the collection in the combine class results in an error.

### Preconditions
Install smile magezon page builder for example
Magento Version : 2.4.3-p1
ElasticSuite Version : 2.10.9.1
Environment : Developer

Third party modules : magezon page builder

### Steps to reproduce
1. Add a Widget to an cms page with bestseller products + rule like show only products with prices greater 10
2. open the cms page => sql error

### Expected result
1. cms page including widget

### Actual result
Exception #0 (Zend_Db_Statement_Exception): SQLSTATE[42S22]: Column not found: 1054 Unknown column 'at_price.value' in 'where clause', query was: 
```
SELECT COUNT(DISTINCT e.entity_id) FROM `catalog_product_entity` AS `e`
 INNER JOIN `catalog_product_entity_int` AS `at_visibility_default` ON (`at_visibility_default`.`entity_id` = `e`.`entity_id`) AND (`at_visibility_default`.`attribute_id` = '102') AND `at_visibility_default`.`store_id` = 0
 LEFT JOIN `catalog_product_entity_int` AS `at_visibility` ON (`at_visibility`.`entity_id` = `e`.`entity_id`) AND (`at_visibility`.`attribute_id` = '102') AND (`at_visibility`.`store_id` = 1)
 INNER JOIN `catalog_product_index_price` AS `price_index` ON price_index.entity_id = e.entity_id AND price_index.customer_group_id = 0 AND price_index.website_id = '1'
 INNER JOIN `catalog_product_website` AS `product_website` ON product_website.product_id = e.entity_id AND product_website.website_id = 1
 INNER JOIN `sales_bestsellers_aggregated_monthly` AS `aggregation` ON e.entity_id = aggregation.product_id AND aggregation.store_id=1 AND qty_ordered >0
 INNER JOIN `cataloginventory_stock_status` AS `stock_status_index` ON e.entity_id = stock_status_index.product_id WHERE (((IF(at_visibility.value_id > 0, at_visibility.value, at_visibility_default.value) = '2') OR (IF(at_visibility.value_id > 0, at_visibility.value, at_visibility_default.value) = '4'))) AND (((IFNULL(`at_price`.`value`, 0) > 10) )) AND (stock_status_index.stock_status = 1)
```

The correct SQL should be like this:
```
SELECT `e`.*, IF(at_visibility.value_id > 0, at_visibility.value, at_visibility_default.value) AS `visibility`, `price_index`.`price`, `price_index`.`tax_class_id`, `price_index`.`final_price`, IF(price_index.tier_price IS NOT NULL, LEAST(price_index.min_price, price_index.tier_price), price_index.min_price) AS `minimal_price`, `price_index`.`min_price`, `price_index`.`max_price`, `price_index`.`tier_price`, `at_price`.`value` AS `price`
FROM `catalog_product_entity` AS `e`
 INNER JOIN `catalog_product_entity_int` AS `at_visibility_default` ON (`at_visibility_default`.`entity_id` = `e`.`entity_id`) AND (`at_visibility_default`.`attribute_id` = '102') AND `at_visibility_default`.`store_id` = 0
 LEFT JOIN `catalog_product_entity_int` AS `at_visibility` ON (`at_visibility`.`entity_id` = `e`.`entity_id`) AND (`at_visibility`.`attribute_id` = '102') AND (`at_visibility`.`store_id` = 1)
 INNER JOIN `catalog_product_index_price` AS `price_index` ON price_index.entity_id = e.entity_id AND price_index.customer_group_id = 0 AND price_index.website_id = '1'
 INNER JOIN `catalog_product_website` AS `product_website` ON product_website.product_id = e.entity_id AND product_website.website_id = 1
 LEFT JOIN `catalog_product_entity_decimal` AS `at_price` ON (`at_price`.`entity_id` = `e`.`entity_id`) AND (`at_price`.`attribute_id` = '75') AND (`at_price`.`store_id` = 0)
WHERE (((IF(at_visibility.value_id > 0, at_visibility.value, at_visibility_default.value) = '2') OR (IF(at_visibility.value_id > 0, at_visibility.value, at_visibility_default.value) = '4'))) AND (((IFNULL(`at_price`.`value`, 0) > 10) ))
```

The last left join is missing. The problems seems to be the Combine Class of the VirtualCategory.
It returns self without collecting the attributes from the child like the magento core class:
```
        // Create this empty function in order to avoid undefined error
        // @see magento/module-catalog-widget/Block/Product/ProductsList.php:351.
        return $this;
```
vs.
```
        foreach ($this->getConditions() as $condition) {
            $condition->collectValidatedAttributes($collection);
        }
        return $this;
```

I am not sure why this is the case, can anybody help me here. I think there is a good reason for that, but I cannot say why ... thx

**Comment by vahonc** [comment]
Hello @template-provider,

We need more details (screenshots) about the configuration of your CMS page and widget which you are using and the errors that you are getting.

This may be due to a conflict between the two modules. In this case, we cannot guarantee correct functionality if two modules work in the same area. So, you should try disabling the Magezon page builder module and check if the problem that you mentioned still exists.

BR,
Vadym

**Comment by template-provider** [comment]
@vahonc 
thxx for the answer, I hope I can give you an example.
The root cause is actually this commit:
https://github.com/Smile-SA/elasticsuite/commit/e89a1db9474c263a37ec5df9448dc8b55c2a4291

The commit message "[Virtual Category] Pagebuilder Compatibility #ESP-286 " tell me, that this commit should make elasticsuite compatible with a pagebuilder (dont know which one).
But with this commit the "collectValidatedAttributes" from magento is overwritten and imho incompatible.

Reverting that commit makes it compatible again and it works. For me is it hard to understand why the "collectValidatedAttributes" was overwritten?
Now every module/extension that rely on the "collectValidatedAttributes" method is now incompatible.
In the case of the pagebuilder, it was adding the prices attribute to the conditions.

Main question for me is: Why was this removed? (Maybe the internal ticket ESP-286 can help here)
```php
    /**
     * Collect validated attributes for Product Collection
     *
     * @param \Magento\Catalog\Model\ResourceModel\Product\Collection $productCollection
     * @return $this
     */
    public function collectValidatedAttributes($productCollection)
    {
        foreach ($this->getConditions() as $condition) {
            $condition->collectValidatedAttributes($productCollection);
        }
        return $this;
    }
```
=>
```php
    /**
     * Collect validated attributes for Product Collection
     *
     * @param \Magento\Catalog\Model\ResourceModel\Product\Collection $productCollection
     * @return $this
     */
    public function collectValidatedAttributes($productCollection)
    {
        foreach ($this->getConditions() as $condition) {
            $condition->collectValidatedAttributes($productCollection);
        }
        return $this;
    }
```
And why would this makes is more compatible?

**Comment by romainruaud** [comment]
@PierreGauthier @Pigau is the guy who wrote this piece of code.

Waiting from his insight.

**Comment by template-provider** [comment]
@PierreGauthier  @Pigau thx ... looking forward

**Comment by jg-development** [comment]
Hi @PierreGauthier 
I am using the magezon page builder and in the textarea it is possible to use a widget with bestseller products.
https://shop.icletta.com/en/ => Scroll down a bit
The commit from you is saying: Pagebuilder compat ESP-286 .... this may work for some pagebuilders, but not for the magezon.

Is the removal of the collected validated attributes necessary? (Indexed or not)
There is no workaround for the pagebuilder, because it is adding the prices as attributes to the collection.
This is removed by elasticsuite and results in an error.

**Comment by romainruaud** [comment]
Hi

Compatibility with third-party modules is out of the scope of our Open Source support policy provided here, there is not much we can do for your issue.

Feel free to extend the class in your project to restore the collection of validated attributes if your pagebuilder module needs to have it.

Regards

**Comment by otueylue** [comment]
Hello @template-provider,

We have the same problem. 
Did you find a solution to the incompatibility? 

Environment: 
Magento Version: 2.4.5-p1 
ElasticSuite Version: Latest 
Magezon Page Builder: Latest

Thanks & Regards

**Comment by template-provider** [comment]
@otueylue 
Solution A: Override the "Smile\ElasticsuiteVirtualCategory\Model\Rule\WidgetCondition\Combine" class and fix the problem of this extension
Solution B: uninstall smile elasticsuite .... works for me

**Comment by otueylue** [comment]
Hello @template-provider,

Can you recommend me another extension that works with Magezon Page Builder fine and you has good experience.

Thanks for your quick respons,
Regards.

**Comment by otueylue** [comment]
Hello, 

It looks like the issues also occurs with Magento 2 core Page Builder. Is a solution to resolve issues planned?

Thanks & Regards

**Comment by PierreGauthier** [comment]
Hello @mamsincl,

I try to reproduce your issue on a environment with the same versions.
I create this widget in a pagebuilder page :
![image](https://github.com/Smile-SA/elasticsuite/assets/4659198/f3d47792-8812-4711-9bcd-34f45bd29d01)

And it works properly on my side, I was not able to have the mysql error. 
But I wonder how did you extract the counting query `SELECT DISTINCT COUNT(DISTINCT e.entity_id) ...` ?
Because with elasticsuite, the counting query should be processed with elasticsearch not mysql :
This `getSize` here: `magento/module-page-builder/view/frontend/templates/catalog/product/widget/content/carousel.phtml:20` should go to this method `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::loadProductCounts`

@otueylue, we cannot provide you a detailed way to make the magezon module compatible with elasticsuite as  it is not opensource. But they probably override the magento widget product collection. You should consider injecting the elasticsuite production collection factory in the concerned widget product list class as we have done for magento native pagebuilder :
https://github.com/Smile-SA/elasticsuite/blob/2.11.x/src/module-elasticsuite-virtual-category/etc/di.xml#L131

To sum up, the bug is not that the attribute join is missing, the bug is you have the filter query in the mysql query while the filter part should be in the elasticsearch query. With elasticsuite, all filtering parts are made in the elasticsearch query, and it is important to keep it that way to make the boost and virtual categories (among other things) works in every contexts (search, navigation, widgets).

I hope these additional details will help you
Regards
